basic/code17.s

```diff
@@ -4,36 +4,62 @@ conint	jsr posint
 	bne gofuc
 	ldx faclo
 	jmp chrgot
-val	jsr len1
-val_str	bne @1
+
+; the "val" function takes a string and turns it into a number by interpreting
+; the ascii digits etc. Except for the problem that a terminator must be
+; supplied by replacing the character beyond the string, VAL is merely a call
+; to floating point input ("finh").
+val	jsr len1        ;get length
+val_str	bne @1      ;return 0 if len=0
 	jmp zerofc
-@1:	ldx txtptr
+@1:	cmp #254        ;check if string is =>255 chars (including null)
+	bne val_strlong ;yes, return 0
+	ldx txtptr      ;save current txtptr
 	ldy txtptr+1
-	stx strng2
-	sty strng2+1
-	ldx index1
-	stx txtptr
-	clc
-	adc index1
-	sta index2
-	ldx index1+1
-	stx txtptr+1
-	bcc val2
-	inx
-val2	stx index2+1
-	lda (index2)
+	phy
+	phx
+	stx index2+1
+	ldx index1      ;save current index1
+	ldy index1+1
+	phy
+	phx
 	pha
-	lda #0
-	sta (index2)
+	inc
```

Comment:
below this `inc`, we can add `beq val_strlong`.  This means we went from 255 to 0.  Or 255 to 256, which means the string is too long to terminate.  No need for cmp here as the Z flag is set if `inc` causes the .A register to become 0.